File: frontend/packages/core/src/machines.ts
/**
 * Multi-machine model. A client (web/desktop/mobile) holds a list of machine
 * profiles — each is one agent-master daemon — and connects to the active one.
 * There is no central hub; the client talks to each daemon directly.
 */

/** One paired daemon. `token` is a secret; store it securely on desktop. */
export interface MachineProfile {
  id: string;
  name: string;
  baseUrl: string;
  token: string;
}

/** Persisted client state: the machine list plus which one is active. */
export interface PersistedMachines {
  machines: MachineProfile[];
  activeId: string | null;
}

/**
 * Storage adapter for the machine list. Web backs this with localStorage;
 * the desktop app backs it with the OS secure store via the DesktopBridge.
 */
export interface MachineStore {
  load(): Promise<PersistedMachines>;
  save(state: PersistedMachines): Promise<void>;
}

/**
 * Contract the Electron preload exposes on `window.agentMaster`. The web app
 * feature-detects it: when present it uses the OS-encrypted secure store and
 * listens for `agentmaster://pair` deep links; otherwise it falls back to
 * localStorage and manual entry.
 */
export interface DesktopBridge {
  readonly isDesktop: true;
  /** OS-encrypted key/value store (Electron safeStorage backed). */
  secureStore: {
    get(key: string): Promise<string | null>;
    set(key: string, value: string): Promise<void>;
    delete(key: string): Promise<void>;
  };
  /**
   * Subscribe to pairing deep links (`agentmaster://pair?url=&token=&name=`).
   * Returns an unsubscribe function.
   */
  onPair(cb: (payload: PairPayload) => void): () => void;
}

export interface PairPayload {
  url: string;
  token: string;
  name?: string;
}

/** The storage key used for the machine list in either backend. */
export const MACHINES_STORAGE_KEY = 'agent-master.machines';

/** Parse an `agentmaster://pair?...` deep link into a PairPayload, or null. */
export function parsePairLink(link: string): PairPayload | null {
  try {
    const u = new URL(link);
    if (u.protocol !== 'agentmaster:') return null;
    const url = u.searchParams.get('url');
    const token = u.searchParams.get('token');
    if (!url || !token) return null;
    const name = u.searchParams.get('name') ?? undefined;
    return { url, token, name };
  } catch {
    return null;
  }
}

/** Best-effort human label for a machine from its base URL (host[:port]). */
export function defaultMachineName(baseUrl: string): string {
  try {
    const u = new URL(baseUrl);
    return u.port ? `${u.hostname}:${u.port}` : u.hostname;
  } catch {
    return baseUrl;
  }
}

File: android/core/src/main/kotlin/com/agentmaster/core/Machines.kt
package com.agentmaster.core

import kotlinx.serialization.Serializable

/**
 * Multi-machine model. A client (web/desktop/mobile) holds a list of machine
 * profiles — each is one agent-master daemon — and connects to the active one.
 * There is no central hub; the client talks to each daemon directly.
 *
 * Mirrors frontend/packages/core/src/machines.ts.
 */

/** One paired daemon. `token` is a secret; store it securely on device. */
@Serializable
data class MachineProfile(
    val id: String,
    val name: String,
    val baseUrl: String,
    val token: String,
)

/** Persisted client state: the machine list plus which one is active. */
@Serializable
data class PersistedMachines(
    val machines: List<MachineProfile> = emptyList(),
    val activeId: String? = null,
)

/**
 * Storage adapter for the machine list. Web backs this with localStorage; the
 * Android app backs it with EncryptedSharedPreferences (see :app).
 */
interface MachineStore {
    suspend fun load(): PersistedMachines
    suspend fun save(state: PersistedMachines)
}

/** Deep-link pairing payload (`agentmaster://pair?url=&token=&name=`). */
data class PairPayload(
    val url: String,
    val token: String,
    val name: String? = null,
)

/** The storage key used for the machine list in either backend. */
const val MACHINES_STORAGE_KEY = "agent-master.machines"

/**
 * Parse an `agentmaster://pair?...` deep link into a [PairPayload], or null.
 *
 * Mirrors parsePairLink in machines.ts: requires the `agentmaster:` scheme and
 * both `url` and `token` query params; `name` is optional.
 */
fun parsePairLink(link: String): PairPayload? =
    try {
        val u = java.net.URI(link)
        // URI keeps the scheme without a trailing colon.
        if (u.scheme != "agentmaster") {
            null
        } else {
            val params = parseQuery(u.rawQuery)
            val url = params["url"]
            val token = params["token"]
            if (url.isNullOrEmpty() || token.isNullOrEmpty()) {
                null
            } else {
                PairPayload(url = url, token = token, name = params["name"])
            }
        }
    } catch (_: Exception) {
        null
    }

/** Best-effort human label for a machine from its base URL (host[:port]). */
fun defaultMachineName(baseUrl: String): String =
    try {
        val u = java.net.URI(baseUrl)
        val host = u.host ?: return baseUrl
        if (u.port != -1) "$host:${u.port}" else host
    } catch (_: Exception) {
        baseUrl
    }

/** Decode a raw query string into a param map with URL-decoded values. */
private fun parseQuery(rawQuery: String?): Map<String, String> {
    if (rawQuery.isNullOrEmpty()) return emptyMap()
    return rawQuery.split("&").mapNotNull { pair ->
        if (pair.isEmpty()) return@mapNotNull null
        val idx = pair.indexOf('=')
        if (idx < 0) {
            decode(pair) to ""
        } else {
            decode(pair.substring(0, idx)) to decode(pair.substring(idx + 1))
        }
    }.toMap()
}

private fun decode(s: String): String =
    try {
        java.net.URLDecoder.decode(s, "UTF-8")
    } catch (_: Exception) {
        s
    }
